overlord/cmdstate/cmdmgr.go

```diff
@@ -46,28 +46,38 @@ func (m *CommandManager) Ensure() error {
 var defaultExecTimeout = 5 * time.Second
 
 func doExec(t *state.Task, tomb *tomb.Tomb) error {
-	var argv []string
-	var tout time.Duration
-
 	st := t.State()
 	st.Lock()
-	err1 := t.Get("argv", &argv)
-	err2 := t.Get("timeout", &tout)
-	st.Unlock()
-	if err1 != nil {
-		return err1
+	defer st.Unlock()
+
+	var ignore bool
+	if err := t.Get("ignore", &ignore); err != nil && err != state.ErrNoState {
+		return err
+	}
+	if ignore {
+		t.Logf("task ignored")
+		return nil
+	}
+
+	var argv []string
+	var tout time.Duration
+	if err := t.Get("argv", &argv); err != nil {
+		return err
 	}
-	if err2 != nil && err2 != state.ErrNoState {
-		return err2
+
+	err := t.Get("timeout", &tout)
+	if err != nil && err != state.ErrNoState {
+		return err
 	}
-	if err2 == state.ErrNoState {
+	if err == state.ErrNoState {
 		tout = defaultExecTimeout
 	}
 
-	if buf, err := osutil.RunAndWait(argv, nil, tout, tomb); err != nil {
-		st.Lock()
+	st.Unlock()
+	buf, err := osutil.RunAndWait(argv, nil, tout, tomb)
+	st.Lock()
+	if err != nil {
 		t.Errorf("# %s\n%s", strings.Join(argv, " "), buf)
```

Comment:
do we need to hold the lock while using t.Errorf? If so then perhaps we should go back to how the code was before where we manually lock, get stuff from state, then unlock and then keep the current ordering of Unlock and Lock here. Mainly, I'm just concerned that mixing and matching manual locks/unlocks and defered locks/unlocks within a single function will lead to confusion and bugs over time.

Reply:
Yes, Errorf needs a lock.

I think I prefer the mixed approach, because in this case it is only osutil.RunAndWait that needs to run while state is unlocked and other than that the lock can be kept for the entire duration of the function; I think unlock+lock here is very isolated and shouldn't lead to bugs (and fwtw when I worked on these changes and did have bugs, unit tests quickly picked them with panics around locking/unlocking state twice). I have added an extra comment in the code about this, perhaps this makes it more clear. 

FWTW this change was driven by a suggestion from @zyga re err check, which was hard to accomplish without lock changes.